crates/telio-wg/src/no_link_detection.rs

```diff
@@ -0,0 +1,524 @@
+use std::{collections::HashMap, time::Duration};
+use tokio::time::Instant;
+
+use telio_crypto::PublicKey;
+use telio_model::{
+    api_config::FeatureNoLinkDetection,
+    mesh::{LinkState, NodeState},
+};
+
+/// Default wireguard keepalive duration
+const WG_KEEPALIVE: Duration = Duration::from_secs(10);
+/// Default configurable rtt value (15 sec)
+const DEFAULT_CONFIGURABLE_RTT: Duration = Duration::from_secs(15);
+/// Keepalive packets size
+const KEEPALIVE_PACKET_SIZE: u64 = 32;
+
+pub struct NoLinkDetection {
+    config: Config,
+    info: HashMap<PublicKey, NodeInfo>,
+}
+
+impl NoLinkDetection {
+    pub fn new(features: Option<FeatureNoLinkDetection>) -> Self {
+        NoLinkDetection {
+            config: Config::from(features),
+            info: HashMap::new(),
+        }
+    }
+
+    pub fn time_since_last_rx(&self, public_key: &PublicKey) -> Option<Duration> {
+        self.info
+            .get(public_key)
+            .map(|n| n.last_rx_timestamp.elapsed())
+    }
+
+    pub fn insert(
+        &mut self,
+        public_key: &PublicKey,
+        rx_bytes: Option<u64>,
+        tx_bytes: Option<u64>,
+        link_state: Option<LinkState>,
+    ) {
+        let info = NodeInfo {
+            last_rx_bytes: rx_bytes.unwrap_or_default(),
+            last_tx_bytes: tx_bytes.unwrap_or_default(),
+            last_link_state: link_state.unwrap_or(LinkState::Down),
+            ..Default::default()
+        };
+        self.info.insert(*public_key, info);
+    }
+
+    pub fn update(
+        &mut self,
+        public_key: &PublicKey,
+        rx_bytes: Option<u64>,
+        tx_bytes: Option<u64>,
+        node_state: NodeState,
+        push: bool,
+    ) -> NoLinkDetectionUpdateResult {
+        // We want to update info only on pull
+        if push {
+            let link_state = self
+                .info
+                .get(public_key)
+                .map(|n| n.last_link_state)
+                .unwrap_or_else(|| {
+                    if matches!(node_state, NodeState::Connected) {
+                        LinkState::Up
+                    } else {
+                        LinkState::Down
+                    }
+                });
+
+            // Return with no notification generated by no link detection
+            // And last known link_state.
+            return self.build_result(false, node_state, link_state);
+        }
+
+        if let Some(old_info) = self.info.get(public_key) {
+            let rtt = self.rtt();
+            let new_info = old_info.compute_new_node_info(rx_bytes, tx_bytes, rtt);
+            let should_notify = self.should_notify(
+                node_state,
+                old_info.last_link_state,
+                new_info.last_link_state,
+            );
+            let link_state = new_info.last_link_state;
+
+            self.info.insert(*public_key, new_info);
```

Comment:
The comment speaks like it shouldn't happen but it feels like a natural cause of calling update() the first time?